libyul/optimiser/MemoryEscalator.cpp

```diff
@@ -0,0 +1,355 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/MemoryEscalator.h>
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/ExpressionJoiner.h>
+#include <libyul/optimiser/ExpressionSplitter.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/AsmData.h>
+#include <libyul/CompilabilityChecker.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+#include <libyul/Object.h>
+#include <libyul/Utilities.h>
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+#include <libsolutil/Visitor.h>
+#include <list>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace {
+class MemoryInitFinder: ASTModifier
+{
+public:
+	static FunctionCall* run(Block& _block)
+	{
+		MemoryInitFinder memoryInitFinder;
+		memoryInitFinder(_block);
+		return memoryInitFinder.m_memoryInit;
+	}
+	using ASTModifier::operator();
+	void operator()(FunctionCall& _functionCall) override
+	{
+		ASTModifier::operator()(_functionCall);
+		if (_functionCall.functionName.name == "memoryinit"_yulstring)
+		{
+			yulAssert(!m_memoryInit, "More than one memory init.");
+			m_memoryInit = &_functionCall;
+		}
+	}
+private:
+	MemoryInitFinder() = default;
+	FunctionCall* m_memoryInit = nullptr;
+};
+
+class FunctionArgumentCollector: ASTModifier
```

Comment:
This whole thing isn't meant to stay - it's just used to exclude reserving space for function arguments, which can't be escalated so far, s.t. "run until stabilized" doesn't start creating huge memory offsets.
It'll either be unnecessary or change, once we can properly handle function arguments.